service-base/src/main/java/org/eclipse/hono/service/metric/DeviceConnectionTimeTracker.java

```diff
@@ -0,0 +1,153 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.service.metric;
+
+import io.vertx.core.Vertx;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+/**
+ *  A class that tracks and reports the device connection time for tenants.
+ */
+public final class DeviceConnectionTimeTracker {
+
+    private static final long DEVICE_CONNECTION_TIME_RECORDING_INTERVAL_IN_MS = TimeUnit.SECONDS.toMillis(10);
+    protected final Logger log = LoggerFactory.getLogger(DeviceConnectionTimeTracker.class);
+    private final AtomicLong noOfDeviceConnections;
+    private final Consumer<Long> recorder;
+    private final AtomicReference<Instant> startInstant = new AtomicReference<>();
+    private final String tenantId;
+    private final long timerId;
+    private final Vertx vertx;
+
+    private DeviceConnectionTimeTracker(
+            final String tenantId, final Vertx vertx,
+            final long noOfDeviceConnections,
+            final Consumer<Long> recorder) {
+        this.tenantId = tenantId;
+        this.noOfDeviceConnections = new AtomicLong(noOfDeviceConnections);
+        this.recorder = recorder;
+        this.startInstant.set(Instant.now());
+        this.vertx = vertx;
+        this.timerId = this.vertx.setPeriodic(DEVICE_CONNECTION_TIME_RECORDING_INTERVAL_IN_MS, id -> report());
+        log.trace("Stared a device connection time tracker for the tenant [{}].", tenantId);
+    }
+
+    /**
+     * Reports the device connection time and then updates the number of device connections with the given value.
+     * 
+     * @param noOfDeviceConnections The number of device connections.
```

Comment:
can this be 0 or negative?

Reply:
This can be 0 but not negative. I have added now a validation to check if it is negative. If it is negative, then it throws an `IllegalArgumentException`.